activerecord/lib/active_record/attribute_methods.rb

```diff
@@ -33,43 +33,35 @@ def self.set_name_cache(name, value)
 
     BLACKLISTED_CLASS_METHODS = %w(private public protected allocate new name parent superclass)
 
-    class GeneratedAttributeMethods < Module #:nodoc:
-      include Mutex_m
-    end
-
     module ClassMethods
-      def inherited(child_class) #:nodoc:
-        child_class.initialize_generated_modules
-        super
+      def attribute_methods_builder_class
+        ActiveRecord::AttributeMethodsBuilder
       end
 
-      def initialize_generated_modules # :nodoc:
-        @generated_attribute_methods = GeneratedAttributeMethods.new
-        @attribute_methods_generated = false
-        include @generated_attribute_methods
-
+      def inherited(child_class) #:nodoc:
+        if (self == Base || abstract_class?) && !child_class.abstract_class?
+          # When a non-abstract class inherits from ActiveRecord::Base or an
+          # abstract class, we duplicate its attributes methods builder and
+          # include it, in order to get any prefixes/suffixes/affixes that have
+          # been defined on it. We use _attribute_methods_builder here to avoid
+          # defining method_missing/respond_to on the inherited builder, which
+          # would be unnecessary since we define them on the duplicated module.
+          builder = _attribute_methods_builder.dup
```

Comment:
Note to self: `dup`ing an included module builder instance can have some very subtle issues. If we've already defined `method_missing` and `respond_to?`, the builder in the closure will be the original module, not the newly-duped one. So if we go on to add new matchers to the builder we duped, they will not be picked up in those methods.

Two ways to solve this:
- don't define those methods on the original module (so it can be included without incurring an extra code path)
- don't include the original module (so it can define the methods and they won't actually affect the code path)

Still working through the implications of both.